adam_core/utils/indexable.py

```diff
@@ -1,174 +1,451 @@
+        """
+        return self._class_index
 
-    def set_index(self, index: Union[str, np.ndarray]):
-        if isinstance(index, str):
-            class_index = getattr(self, index)
-            self._index_attribute = index
-            array_index = np.arange(0, len(class_index), dtype=int)
+    @index.setter
+    def index(self, value):
+        self.set_index(value)
 
-        elif isinstance(index, np.ndarray):
-            class_index = index
-            self._index_attribute = None
-            array_index = np.arange(0, len(index), dtype=int)
+    @index.deleter
+    def index(self):
+        self.set_index()
+
+    def _check_member_validity(self):
+        """
+        Scans the class's sliceable members and raises an error if they are
+        not all of the same length.
+
+        Sets
+        ----
+        self._member_length : int
+            The length of the members of the class.
+        self._member_index : `~numpy.ndarray`
+            The index of the members of the class.
+
+        Raises
+        ------
+        ValueError: If the members are not all of the same length.
+        """
+        # Scan members and if they are of a type that is sliceable
+        # then store the length. All sliceable members should have the same length
+        # along their first axis.
+        member_lengths = []
+        for k, v in self.__dict__.items():
+            if isinstance(v, SLICEABLE_DATA_STRUCTURES):
+                member_lengths.append(len(v))
 
+        if len(np.unique(member_lengths)) != 1:
+            raise ValueError("All sliceable members must have the same length.")
         else:
-            err = "index must be a str, numpy.ndarray"
-            raise ValueError(err)
+            member_length = member_lengths[0]
 
-        self._index = pd.MultiIndex.from_arrays(
-            [class_index, array_index], names=["class_index", "array_index"]
-        )
+        self._member_length = member_length
+        self._member_index = np.arange(0, member_length, dtype=int)
 
         return
 
-    def __getitem__(self, i: Union[int, slice, tuple, list, np.ndarray, pd.MultiIndex]):
-        ind = self._handle_index(i)
-        copy = deepcopy(self)
+    def set_index(self, index_values: Optional[Union[str, npt.ArrayLike]] = None):
+        """
+        Set an index on the class for the given values or attribute name.
 
-        for k, v in self.__dict__.items():
-            if isinstance(
-                v,
-                (np.ndarray, np.ma.masked_array, list, Time, Indexable, pd.MultiIndex),
-            ):
-                copy.__dict__[k] = v[ind]
-            elif isinstance(v, UNSLICEABLE_DATA_STRUCTURES):
-                copy.__dict__[k] = v
-            elif v is None:
-                pass
-            else:
-                err = f"{type(v)} are not supported."
-                raise NotImplementedError(err)
+        Parameters
+        ----------
+        values : str, np.ArrayLike
```

Comment:
The name here is out of sync with the function signature.

Reply:
Fixed.